homeassistant/components/folder_watcher.py

```diff
@@ -0,0 +1,96 @@
+"""
+Component for monitoring activity on a folder.
+"""
+import os
+import logging
+import voluptuous as vol
+from homeassistant.helpers.entity import Entity
+from homeassistant.const import EVENT_HOMEASSISTANT_START
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['watchdog==0.8.3']
+_LOGGER = logging.getLogger(__name__)
+
+CONF_FOLDERS = 'folders'
+CONF_FILTERS = 'filters'
+DEFAULT_FILTER = '*'
+DOMAIN = "folder_watcher"
+EVENT_TYPE = "event_type"
+FILE = 'file'
+FOLDER = 'folder'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_FOLDERS):
+            vol.All(cv.ensure_list, [cv.isdir]),
+        vol.Optional(CONF_FILTERS, default=[DEFAULT_FILTER]):
+            vol.All(cv.ensure_list, [cv.string]),
+        }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the folder watcher."""
+    conf = config[DOMAIN]
+    paths = conf[CONF_FOLDERS]
+    patterns = conf[CONF_FILTERS]
+
+    def run_setup(event):
+        for path in paths:
+            if not hass.config.is_allowed_path(path):
+                _LOGGER.error("folder %s is not valid or allowed", path)
+                return False
+            else:
+                Watcher(path, patterns, hass)
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, run_setup)
+    return True
+
+
+def create_event_handler(patterns, hass):
+    from watchdog.events import PatternMatchingEventHandler
+
+    class EventHandler(PatternMatchingEventHandler):
+        """Class for handling Watcher events."""
+
+        def __init__(self, patterns, hass):
+            super().__init__(patterns)
+            self.hass = hass
+
+        def process(self, event):
+            """On Watcher event, fire HA event."""
+            if not event.is_directory:
+                file_name = os.path.split(event.src_path)[1]
+                folder_name = os.path.split(event.src_path)[0]
+                self.hass.bus.fire(
+                    DOMAIN, {
+                        EVENT_TYPE: event.event_type,
+                        FILE: file_name,
+                        FOLDER: folder_name
+                        })
+
+        def on_modified(self, event):
+            self.process(event)
+
+        def on_moved(self, event):
+            self.process(event)
+
+        def on_created(self, event):
+            self.process(event)
+
+        def on_deleted(self, event):
+            self.process(event)
+
+    return EventHandler(patterns, hass)
+
+
+class Watcher(Entity):
+    """Class for starting Watchdog."""
+    def __init__(self, path, patterns, hass):
+        from watchdog.observers import Observer
+        self._observer = Observer()
+        self._observer.schedule(
+            create_event_handler(patterns, hass),
+            path,
+            recursive=True)
+        self._observer.start()
```

Comment:
If I understand correctly, this will create a new thread (`threading.Thread`). We need to stop that thread when shutting down Home Assistant, because otherwise we would significantly slow down shut down.

For example, the cast integration did the same for a long time. It created a new thread for each discovered cast and never closed them on Home Assistant shutdown. This made Home Assistant hang for about 10s before shutting down after calling `homeassistant.stop`.

So you should additionally listen to `EVENT_HOMEASSISTANT_STOP` and clean up the threads there with something like

```
observer.stop()
# maybe also call "observer.join()", though that may not 
# actually be necessary because we don't care about the result.
```

Reply:
Oh, it seems like you had this in another commit, but then reverted it... Was there a reason/a discussion for that I missed? Is `watchdog` handling this for us already?

Reply:
OK this appears to work:
```python
class Watcher(Entity):
    """Class for starting Watchdog."""
    def __init__(self, path, patterns, hass):
        from watchdog.observers import Observer
        self._observer = Observer()
        self._observer.schedule(
            create_event_handler(patterns, hass),
            path,
            recursive=True)
        self._observer.start()
        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)

    def shutdown(self, event):
        """Shutdown the watcher."""
        self._observer.stop()
```